### PR TITLE
draft: dpdk: Introduce support for DPDK secondary mode for workers runmode

### PR DESCRIPTION
**Pull request by lukashino**
Implement secondary mode operation in Suricata. Receive
packets over DPDK rings. To each worker assign a separate ring.
Provide configuration options specific for the secondary mode
along with possibility to specify arbitrary ring name.
Change threading from threads to DPDK lcores.
Add relevant documentation.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/4910) ticket:

### CODE REVIEW

File: doc/userguide/conf.py [hunk 1]
```diff
@@ -335,3 +335,7 @@ def setup(app):
     "sysconfdir": os.getenv("sysconfdir", "/etc"),
     "localstatedir": os.getenv("localstatedir", "/var"),
 }
+
+
+# for using reference numbers
+numfig = True
```
Comment (victorjulien) [inline doc/userguide/conf.py:341]: this should go in a separate commit

File: doc/userguide/configuration/suricata-yaml.rst [hunk 2]
```diff
@@ -2349,7 +2439,7 @@ The Teredo decoder can be disabled. It is enabled by default.
         ports: $TEREDO_PORTS # syntax: '[3544, 1234]'
 
 Using this default configuration, Teredo detection will run on UDP port
-3544. If the `ports` parameter is missing, or set to `any`, all ports will be
+1.    If the `ports` parameter is missing, or set to `any`, all ports will be
```
Comment (victorjulien) [inline doc/userguide/configuration/suricata-yaml.rst:2442]: this seems unintended

File: src/runmode-dpdk.c [hunk 3]
```diff
@@ -1102,6 +1130,87 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
+    if (q_num > 9999)
+        FatalError(SC_ERR_DPDK_CONF, "Number of queues must not exceed 10000 queues per interface");
+    q_pos = strstr(format, DPDK_CONFIG_DEFAULT_QUEUE_NUM_SPECIFIER);
+    snprintf(name, q_pos - format + 1, "%s", format);
+    snprintf(q_num_str, sizeof(q_num_str), "%" PRIu16, q_num);
+    strlcat(name, q_num_str, strlen(q_num_str));
+    // copy the rest after queue number specifier
+    strlcat(name, q_pos + strlen(DPDK_CONFIG_DEFAULT_QUEUE_NUM_SPECIFIER),
+            sizeof(name) - strlen(name) - 1);
+    strlcpy(name_out, name, sizeof(name));
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1148]: the size arg needs to be the size of the target buffer, which should be passed as an arg to this function

File: src/runmode-dpdk.c [hunk 4]
```diff
@@ -1102,6 +1130,87 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
+}
+
+static bool DeviceRingNameIsValid(const char *name)
+{
+    unsigned long len = strlen(name);
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1153]: use explicit types like `uint32_t`

File: src/runmode-dpdk.c [hunk 5]
```diff
@@ -1102,6 +1130,87 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
+        SCReturnInt(-ENOMEM);
+    }
+
+    if (!DeviceRingNameIsValid(iconf->iface))
+        SCReturnInt(-EINVAL);
+
+    for (int32_t i = 0; i < iconf->threads; ++i) {
+        char name[RTE_RING_NAMESIZE];
+        DeviceRingNameInit(name, iconf->iface, i);
+        SCLogInfo("Looking up rx ring: %s", name);
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1188]: Debug

File: src/runmode-dpdk.c [hunk 6]
```diff
@@ -1102,6 +1130,87 @@ static int DeviceConfigureIPS(DPDKIfaceConfig *iconf)
...
+        iconf->rx_rings[i] = rte_ring_lookup(name);
+        if (iconf->rx_rings[i] == NULL) {
+            SCLogError(SC_ERR_DPDK_INIT, "rte_ring_lookup(): cannot get rx ring '%s'\n", name);
```
Comment (victorjulien) [inline src/runmode-dpdk.c:1191]: no `\n`

Comment (victorjulien) [reply src/runmode-dpdk.c:1191]: multiple instances of this issue, please review all

File: src/tm-threads.c [hunk 7]
```diff
@@ -248,7 +283,7 @@ static void *TmThreadsSlotPktAcqLoop(void *td)
                                  " tmqh_out=%p",
                    s, s ? s->PktAcqLoop : NULL, tv->tmqh_in, tv->tmqh_out);
         TmThreadsSetFlag(tv, THV_CLOSED | THV_RUNNING_DONE);
-        pthread_exit((void *) -1);
+        TmThreadExit(tv, -1);
```
Comment (victorjulien) [inline src/tm-threads.c:286]: Initial change should go into its own commit, then add the specific dpdk changes

File: src/util-dpdk.c [hunk 8]
```diff
@@ -18,16 +18,22 @@
...
  */
 
+#ifndef UTIL_DPDK_C
+#define UTIL_DPDK_C
+
 #include "suricata.h"
 #include "util-dpdk.h"
 
+// equal to -1 to get the first lcore id
+int dpdk_last_spawned_lcore = -1;
```
Comment (victorjulien) [inline src/util-dpdk.c:31]: how/why is this global used?

**Review by victorjulien — COMMENTED** [review]
Incomplete review with some code comments.

**Comment by victorjulien** [comment]
Is prefilter mode the same as secondary mode? Or are these separate concepts?

**Comment by lukashino** [comment]
> Is prefilter mode the same as secondary mode? Or are these separate concepts?

They are pretty much the same but I have not mentioned prefilter in the docs. So maybe it makes more sense to call it secondary mode on Suricata side and "prefilter" should be called only the application that is in front of Suricata.

File: doc/userguide/conf.py (hunk 1 above)
Comment (lukashino) [reply doc/userguide/conf.py:341]: Should only this change (regarding conf.py change) be in a separate commit or all the docs changes?

File: doc/userguide/configuration/suricata-yaml.rst (hunk 2 above)
Comment (lukashino) [reply doc/userguide/configuration/suricata-yaml.rst:2442]: thanks

File: src/runmode-dpdk.c (hunk 6 above)
Comment (lukashino) [reply src/runmode-dpdk.c:1191]: should be fixed

File: doc/userguide/conf.py (hunk 1 above)
Comment (victorjulien) [reply doc/userguide/conf.py:341]: this updated general doc config, so that should be separate from any dpdk specific thing

**Comment by victorjulien** [comment]
> > Is prefilter mode the same as secondary mode? Or are these separate concepts?
> 
> They are pretty much the same but I have not mentioned prefilter in the docs. So maybe it makes more sense to call it secondary mode on Suricata side and "prefilter" should be called only the application that is in front of Suricata.

I see several refs to prefilter in code (and config?). Would like to avoid confusion. I think for this phase we should limit to secondary?

File: src/tm-threads.c (hunk 7 above)
Comment (lukashino) [reply src/tm-threads.c:286]: done

File: src/runmode-dpdk.c (hunk 4 above)
Comment (lukashino) [reply src/runmode-dpdk.c:1153]: I've done my best to refactor this in other places as well.

File: doc/userguide/conf.py (hunk 1 above)
Comment (lukashino) [reply doc/userguide/conf.py:341]: ok, the change has been separated.

File: src/runmode-dpdk.c (hunk 3 above)
Comment (lukashino) [reply src/runmode-dpdk.c:1148]: Argument has been added to the function.

File: src/util-dpdk.c (hunk 8 above)
Comment (lukashino) [reply src/util-dpdk.c:31]: The global variable was used to store the last lcore onto which a Suricata worker/management thread had been spawned. However, as I am only spawning workers in the current PR, I've moved the variable definition inside of the function `RunModeSetLiveCaptureWorkersForDevice`.

**Comment by lukashino** [comment]
Opening a new one at https://github.com/OISF/suricata/pull/7092
